### PR TITLE
Refactor smithy-rules-engine

### PR DESCRIPTION
**Pull request by kstich**
Remaining tasks:
- [x] JavaDoc pass
- [x] Smithy specification

-----

[Significant refactoring of smithy-rules-engine](https://github.com/awslabs/smithy/commit/bc01b57c401e8c129bcbdd96827951d44ce05bde) 

This commit collects many changes to limit the public interface,
align with Java standards, align with idioms from the rest of the
Smithy codebase, and generally clean up the codebase.

Several packages, classes, and methods have been removed, renamed,
moved, or changed visibility. Use of `assert` has been removed and
the usage of `stream` has been reduced. Classes with multiple subtypes
have had their subtype names adjusted for compatibility with Java and
been broken out to their own packages and classes.

-----

[Enable RuleSet and Test validation by default](https://github.com/awslabs/smithy/commit/d062610f862fb57598f557542990ce31392aadf0) 

This commit removes the custom validation system that required code
to validate a ruleset and its tests, integrating the validation with
Smithy's built in ModelValidation system. Several bugs in validators
and test cases were fixed. Test cases were moved to using the standard
errorfiles based setup where applicable.

Some related refactoring was also performed.

-----

[Add validation of endpoint test operation inputs](https://github.com/awslabs/smithy/commit/f485a6bbaabfa3e022d60304207d6b15f25a621e) 

This commit adds validation that for every endpoint test case that
contains operationInputs, the operation input entries match their
specified operation's input structure.

Co-authored-by: Russell Cohen <rcoh@amazon.com>

-----

[Continued refactoring of smithy-rules-engine](https://github.com/awslabs/smithy/commit/a74c8d8dfa08f01a64dfe17dce307b115113a747) 
This commit collects many changes to limit the public interface,
align with Java standards, align with idioms from the rest of the
Smithy codebase, and generally clean up the codebase.

Several packages, classes, and methods have been removed, renamed,
moved, or changed visibility. Direct access to inner mutable types
has been removed. The internal complexity of the CoverageChecker
has been significantly reduced while maintaining used functionality.

-----

[Refactor rules engine functions](https://github.com/awslabs/smithy/commit/8fc4e78e2b4e664057da1acb7dc1fb328389c40b) 

This commit refactors the configuration and inheritance hierarchy
of functions in the rules engine. Two intermediate abstract classes
have been eliminated, SingleArgFunction and Function.

All functions are now supplied through an SPI on FunctionDefinition,
and each implementation now supplies a FunctionDefinition. Special
cases for methods when generating from FunctionNode instances have
also been removed, aligning all implementations to the same
construction. No behavior has changed in the ExpressionVisitor to
reduce the impact of migration.

Unused, or minimally test-case used, convenience functions have been
removed.

Several bugs were fixed, including dropping context information from
certain sets of RulesError generation.

-----

[Extract AWS specific rules engine functions](https://github.com/awslabs/smithy/commit/78dda8d26f56253f23823e16867d7cdcbfd61a73) 

This commit pulls the AWS specific rules engine functions into a new
package, smithy-aws-rules-engine. This includes parseArn, partition,
and isVirtualHostableS3Bucket. The partition loading has also moved
to this package, including its data types.

Relevant tests were migrated. Behavior that was only possible to
validate using the parseArn function have been moved, even if the
behavior is not AWS specific. Tests that used AWS functions but did
not require it were updated to be agnostic.

-----

[Update license headers](https://github.com/awslabs/smithy/commit/d54fcd75644e8fd32c99c88d814f5d319ae0acfb)

-----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: smithy-aws-rules-engine/src/test/resources/software/amazon/smithy/aws/rulesengine/language/functions/errorfiles/invalid/eventbridge-rules-extra-param.errors [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[WARNING] example#FizzBuzz: This shape applies a trait that is unstable: smithy.rules#endpointRuleSet | UnstableTrait
+[WARNING] example#FizzBuzz: This shape applies a trait that is unstable: smithy.rules#endpointTests | UnstableTrait
+[WARNING] example#FizzBuzz: This shape applies a trait that is unstable: smithy.rules#clientContextParams | UnstableTrait
+[DANGER] example#FizzBuzz: Parameter 'useDualStackEndpoint' is never used in test cases | RuleSetParameter.Unused
```
Comment (JordonPhillips) [inline smithy-aws-rules-engine/src/test/resources/software/amazon/smithy/aws/rulesengine/language/functions/errorfiles/invalid/eventbridge-rules-extra-param.errors:4]: This is really confusing - it needs to reference the trait that the issue is under. Otherwise you could could reasonably think it's referring to the endpoint rules trait.

File: smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/language/Endpoint.java [hunk 2]
```diff
@@ -1,91 +1,75 @@
...
  */
 @SmithyUnstableApi
-public final class Endpoint extends MandatorySourceLocation implements ToSmithyBuilder<Endpoint>, TypeCheck, ToNode {
+public final class Endpoint implements FromSourceLocation, ToNode, ToSmithyBuilder<Endpoint>, TypeCheck {
     private static final String URL = "url";
     private static final String PROPERTIES = "properties";
     private static final String HEADERS = "headers";
     private static final String SIGV_4 = "sigv4";
     private static final String SIG_V4A = "sigv4a";
     private static final String SIGNING_REGION = "signingRegion";
```
Comment (JordonPhillips) [inline smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/language/Endpoint.java:48]: Why are these here? Could this just be a TypedPropertiesBag or have some other method of extending what properties an endpoint can have

File: smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/validators/RuleSetAuthSchemesValidator.java
Comment (JordonPhillips) [inline smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/validators/RuleSetAuthSchemesValidator.java:1]: All of these validators need to have more context: the trait and ideally a node path to the place where the error is.

File: smithy-rules-engine/src/test/resources/software/amazon/smithy/rulesengine/traits/errorfiles/missing-required-values.smithy [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+namespace smithy.example
+
+use smithy.rules#endpointTests
+
+service InvalidService {
+    version: "2022-01-01",
+    operations: [GetThing]
+}
+
+apply InvalidService @endpointTests({
```
Comment (JordonPhillips) [inline smithy-rules-engine/src/test/resources/software/amazon/smithy/rulesengine/traits/errorfiles/missing-required-values.smithy:12]: looks like we need to fix our syntax highlighter

File: smithy-utils/src/main/java/software/amazon/smithy/utils/StringUtils.java [hunk 4]
```diff
@@ -1605,4 +1605,21 @@ public static int levenshteinDistance(CharSequence left, CharSequence right, fin
...
+     * @return the indented string.
+     */
+    public static String indent(String string, int spaces) {
+        StringBuilder indentBuilder = new StringBuilder(spaces + 2);
+        for (int i = 0; i < spaces; ++i) {
+            indentBuilder.append(" ");
+        }
+        String indent = indentBuilder.toString();
+
+        return indent + string.replace("\n", "\n" + indent);
```
Comment (JordonPhillips) [inline smithy-utils/src/main/java/software/amazon/smithy/utils/StringUtils.java:1623]: rstrip

File: smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/language/RulesComponentBuilder.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+     *
+     * @param e StackTraceElement to test.
+     * @return Returns true if this element should be in a comment.
+     */
+    static boolean isStackTraceRelevant(StackTraceElement e) {
+        String normalized = e.getClassName().replace("$", ".");
+        return !normalized.startsWith("java.")
+                       && !normalized.startsWith("jdk.")
+                       && !normalized.startsWith(RulesComponentBuilder.class.getCanonicalName())
+                       && !normalized.startsWith("software.amazon.smithy.rulesengine");
```
[the remark this answers is not in the document]
Comment (kstich) [reply smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/language/RulesComponentBuilder.java:80]: Resolved by adjusting the namespace of the AWS package components.

File: smithy-aws-rules-engine/src/main/java/software/amazon/smithy/rulesengine/aws/language/functions/ParseArn.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                    SERVICE, Type.stringType(),
+                    REGION, Type.stringType(),
+                    ACCOUNT_ID, Type.stringType(),
+                    RESOURCE_ID, Type.arrayType(Type.stringType())
+            )));
+        }
+
+        @Override
+        public Value evaluate(List<Value> arguments) {
+            String value = arguments.get(0).expectStringValue().getValue();
```
[the remark this answers is not in the document]
Comment (kstich) [reply smithy-aws-rules-engine/src/main/java/software/amazon/smithy/rulesengine/aws/language/functions/ParseArn.java:73]: This is done as part of the `typeCheckLocal` call, via `checkTypeSignature` and the definition's `getArguments` method.

File: smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/analysis/CoverageChecker.java [hunk 7]
```diff
@@ -46,7 +32,7 @@ public final class CoverageChecker {
 
     public CoverageChecker(EndpointRuleSet ruleSet) {
         this.ruleSet = ruleSet;
-        this.checkerCore = new CoverageCheckerCore();
+        checkerCore = new CoverageCheckerCore();
```
[the remark this answers is not in the document]
Comment (kstich) [reply smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/analysis/CoverageChecker.java:35]: Wasn't necessary, as there's no other `checkerCore` in scope like there is for `ruleSet`.

File: smithy-aws-rules-engine/src/main/java/software/amazon/smithy/rulesengine/aws/language/functions/AwsPartition.java [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+        return visitor.visitLibraryFunction(DEFINITION, getArguments());
+    }
+
+    private static PartitionData loadPartitionData() {
+        Iterator<PartitionDataProvider> iter = ServiceLoader.load(PartitionDataProvider.class).iterator();
+        if (!iter.hasNext()) {
+            throw new RuntimeException("Unable to locate partition data");
+        }
+
+        PartitionDataProvider provider = iter.next();
```
[the remark this answers is not in the document]
Comment (kstich) [reply smithy-aws-rules-engine/src/main/java/software/amazon/smithy/rulesengine/aws/language/functions/AwsPartition.java:62]: I don't think we'd want to throw, as there's no non-code mechanism to override this. I'll explore what can be changed here.

File: smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/language/syntax/expressions/functions/ParseUrl.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+        @Override
+        public List<Type> getArguments() {
+            return Collections.singletonList(Type.stringType());
+        }
+
+        @Override
+        public Type getReturnType() {
+            Map<Identifier, Type> types = new LinkedHashMap<>();
```
Comment (hpmellema) [inline smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/language/syntax/expressions/functions/ParseUrl.java:70]: nit: another place that could be statically initialized.

File: smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/language/syntax/expressions/functions/Substring.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        @Override
+        public Value evaluate(List<Value> arguments) {
+            String str = arguments.get(0).expectStringValue().getValue();
+            int startIndex = arguments.get(1).expectIntegerValue().getValue();
+            int stopIndex = arguments.get(2).expectIntegerValue().getValue();
+            boolean reverse = arguments.get(3).expectBooleanValue().getValue();
+
+            for (int i = 0; i < str.length(); i++) {
+                char ch = str.charAt(i);
+                if (!(ch <= 127)) {
```
Comment (hpmellema) [inline smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/language/syntax/expressions/functions/Substring.java:71]: why 127?

File: smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/validators/RuleSetAuthSchemesValidator.java [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+    private class Validator extends TraversingVisitor<ValidationEvent> {
+        private final ServiceShape serviceShape;
+
+        Validator(ServiceShape serviceShape) {
+            this.serviceShape = serviceShape;
+        }
+
+        @Override
+        public Stream<ValidationEvent> visitEndpoint(Endpoint endpoint) {
```
Comment (hpmellema) [inline smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/validators/RuleSetAuthSchemesValidator.java:55]: question: why return the Stream here instead of just the list? It looks like it just gets collected to a list anyway?

File: smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/validators/RuleSetParamMissingDocsValidator.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        }
+        return events;
+    }
+
+    public List<ValidationEvent> validateRuleSet(ServiceShape serviceShape, EndpointRuleSet ruleSet) {
+        List<ValidationEvent> events = new ArrayList<>();
+        for (Parameter parameter : ruleSet.getParameters()) {
+            if (!parameter.getDocumentation().isPresent()) {
+                events.add(warning(serviceShape, parameter,
+                        String.format("Parameter `%s` did not have documentation", parameter.getName())));
```
Comment (hpmellema) [inline smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/validators/RuleSetParamMissingDocsValidator.java:37]:
```suggestion
                        String.format("Parameter `%s` does not have documentation", parameter.getName())));
```

File: smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/language/syntax/expressions/Expression.java [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+     *
+     * @param node the node to construct the expression from.
+     * @return the expression.
+     */
+    public static Expression fromNode(Node node) {
+        if (node.asObjectNode().isPresent()) {
+            ObjectNode on = node.asObjectNode().get();
+            Optional<Node> ref = on.getMember("ref");
+            Optional<Node> fn = on.getMember("fn");
+            if ((ref.isPresent() ? 1 : 0) + (fn.isPresent() ? 1 : 0) != 1) {
```
[the remark this answers is not in the document]
Comment (hpmellema) [reply smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/language/syntax/expressions/Expression.java:78]: I see now what this is doing.  I think a quick comment here would be helpful that this is a mutually exclusive check.

[the remark this answers is not in the document]
Comment (kstich) [reply smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/language/syntax/expressions/Expression.java:78]: No, one and only one is allowed at a time.

File: smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/language/syntax/expressions/functions/Substring.java (hunk 10 above)
Comment (kstich) [reply smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/language/syntax/expressions/functions/Substring.java:71]: Only standard ASCII chars are supported by the rules engine. This will be detailed in specification.

File: smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/validators/RuleSetAuthSchemesValidator.java (hunk 11 above)
Comment (kstich) [reply smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/validators/RuleSetAuthSchemesValidator.java:55]: `TraversingVisitor` runs on `Stream`s of the parameterized type.

**Review by hpmellema — APPROVED** [review]

**Review by JordonPhillips — APPROVED** [review]

File: docs/source-2.0/additional-specs/rules-engine/parameters.rst [hunk 14]
```diff
@@ -0,0 +1,334 @@
+.. _rules-engine-parameters:
+
+=======================
+Rules engine parameters
+=======================
+
+:ref:`Rule set parameters <rules-engine-endpoint-rule-set-parameter>` are a set
```
Comment (hpmellema) [inline docs/source-2.0/additional-specs/rules-engine/parameters.rst:7]: Nit: the title of this page is "Rules Engine Parameters" but they are call "Rule set parameters" here?

File: docs/source-2.0/additional-specs/rules-engine/parameters.rst [hunk 15]
```diff
@@ -0,0 +1,334 @@
...
+
+Implementations MUST always generate a parameter structure, even in the absence
+of parameters. This provides for a stable API interface regardless of a
+service’s rule set evolution over time. The addition of parameters to a rule
+set MUST be a backwards-compatible.
+
+Implementations MUST generate parameter types using their language idiomatic
+interfaces to distinguish whether values are set or unset by users. For
+example, the Go programming language would generate these types as pointer
+values, where Rust would wrap types using the ``Option`` type.
```
Comment (hpmellema) [inline docs/source-2.0/additional-specs/rules-engine/parameters.rst:60]:
```suggestion
values, whereas Rust would wrap types using the ``Option`` type.
```

File: docs/source-2.0/additional-specs/rules-engine/parameters.rst [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+
+.. _rules-engine-parameters-binding-values:
+
+------------------------
+Binding parameter values
+------------------------
+
+Rules engine implementations bind values to parameters defined in rule sets,
+utilizing information in the service model in addition to the rules.
+`Rules engine parameter traits`_ are used bind runtime values to rule sets.
```
Comment (hpmellema) [inline docs/source-2.0/additional-specs/rules-engine/parameters.rst:86]:
```suggestion
`Rules engine parameter traits`_ are used as bind runtime values to rule sets.
```

File: docs/source-2.0/additional-specs/rules-engine/index.rst [hunk 17]
```diff
@@ -0,0 +1,17 @@
+.. _rules-engine:
+
+============
+Rules engine
+============
+
+The Smithy rules engine provides service owners with a collection of traits
+and components to define rule sets. Rule sets can be consumed by code
+generators to provide the rule set's specific functionality, e.g. endpoint
+resolution.
```
Comment (hpmellema) [inline docs/source-2.0/additional-specs/rules-engine/index.rst:10]: Any other examples other than just endpoint resolution? Might help clarify

File: docs/source-2.0/aws/rules-engine/library-functions.rst [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+    .. list-table::
+        :header-rows: 1
+        :widths: 10 20 70
+
+        * - Property
+          - Type
+          - Description
+        * - partition
+          - ``string``
+          - The partition where the resource is located.
```
Comment (hpmellema) [inline docs/source-2.0/aws/rules-engine/library-functions.rst:131]: Should we link to https://docs.aws.amazon.com/whitepapers/latest/aws-fault-isolation-boundaries/partitions.html?

File: docs/source-2.0/aws/rules-engine/library-functions.rst [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+    }
+
+
+.. _rules-engine-aws-library-awsPartition-Partition:
+
+-----------------------
+``Partition`` structure
+-----------------------
+
+The ``Partition`` structure is returned from the `aws.partition function`_ when
```
Comment (hpmellema) [inline docs/source-2.0/aws/rules-engine/library-functions.rst:57]: Maybe also link this to https://docs.aws.amazon.com/whitepapers/latest/aws-fault-isolation-boundaries/partitions.html

File: docs/source-2.0/additional-specs/rules-engine/specification.rst [hunk 20]
```diff
@@ -0,0 +1,688 @@
+.. _rules-engine-specification:
+
+==========================
+Rules engine specification
+==========================
+
+The Smithy rules engine provides service owners with a collection of traits
+and components to define rule sets. Rule sets can be consumed by code
```
Comment (hpmellema) [inline docs/source-2.0/additional-specs/rules-engine/specification.rst:8]: nit: link to rule set page

File: docs/source-2.0/additional-specs/rules-engine/specification.rst [hunk 21]
```diff
@@ -0,0 +1,688 @@
...
+                    }
+                ]
+            }
+        ],
+        "type": "tree",
+        "rules": [
+            {
+                "type": "tree",
+                "conditions": [
+                    // Abbreviated for clarity
```
Comment (hpmellema) [inline docs/source-2.0/additional-specs/rules-engine/specification.rst:444]: nit: Above you just have `/* abbreviated */`. Should probably standardize that.

**Review by hpmellema — CHANGES_REQUESTED** [review]

**Review by mtdowling — CHANGES_REQUESTED** [review]
Just some quick notes, need to sit down with this for ... a while to fully review it :)

**Review by hpmellema — APPROVED** [review]

File: docs/source-2.0/additional-specs/rules-engine/standard-library.rst [hunk 22]
```diff
@@ -0,0 +1,453 @@
...
+      - ``string``
+      - The URL scheme, such as ``https``. The value returned MUST NOT
+        include the ``://`` separator.
+    * - authority
+      - ``string``
+      - The host and optional port component of the URL. A default port
+        MUST NOT be included. A userinfo segment MUST NOT be included.
+    * - path
+      - ``string``
+      - The unmodified, path segment of the URL.
```
Comment (JordonPhillips) [inline docs/source-2.0/additional-specs/rules-engine/standard-library.rst:265]:
```suggestion
      - The unmodified path segment of the URL.
```

File: docs/source-2.0/additional-specs/rules-engine/standard-library.rst [hunk 23]
```diff
@@ -0,0 +1,453 @@
...
+===================================
+
+Extensions to the rules engine can provide additional functions. Code
+generators MAY support these additional functions and SHOULD document which
+extensions are supported. Additional functions MUST be namespaced, using
+two colon ``:`` characters to separate namespace portions. This is utilized to
+add the :ref:`AWS rules engine functions <rules-engine-aws-library-functions>`.
+
+The rules engine is highly extensible through
+``software.amazon.smithy.rulesengine.language.EndpointRuleSetExtension``
```
Comment (JordonPhillips) [inline docs/source-2.0/additional-specs/rules-engine/standard-library.rst:450]: You should link to the javadoc here

**Review by JordonPhillips — APPROVED** [review]

**Review by JordonPhillips — APPROVED** [review]

**Comment by kstich** [comment]
Moved to #1855 using a branch on the main repository instead of a fork.
